torch/csrc/distributed/autograd/engine/dist_engine.cpp

```diff
@@ -23,6 +23,53 @@ using torch::autograd::variable_list;
 static constexpr char* kNumBackwardPasses = "num_current_backward_passes";
 static constexpr char* kNumAutogradContexts = "num_autograd_contexts";
 
+// This hook does 3 things:
+//   1. Accumuate the gard to RPC context.
+//   2. Call pre hooks of the original AccumulateGrad to modify the input grad.
+//   3. Call post hooks of the original AccumulateGrad.
+class DistAccumulateGradCaptureHook
+    : public GraphTask::ExecInfo::Capture::GradCaptureHook {
+ public:
+  DistAccumulateGradCaptureHook(
+      std::shared_ptr<AccumulateGrad> accumulateGrad,
+      ContextPtr autogradContext)
+      : accumulateGrad_(std::move(accumulateGrad)),
+        autogradContext_(std::move(autogradContext)) {}
+
+  at::Tensor operator()(const at::Tensor& grad) override {
+    variable_list inputGrads = {grad};
+    // It's intended that pre/post hooks are still called even if the grad is
+    // undenfined here.
+    for (const auto& hook : accumulateGrad_->pre_hooks()) {
+      inputGrads = (*hook)(inputGrads);
+    }
+
+    // It is possible that the grad is not defined since a separate
+    // invocation of the autograd engine on the same node might actually
+    // compute this gradient.
+    if (inputGrads[0].defined()) {
+      // Three internal references to 'inputGrads[0]' at this moment:
+      //   1. 'inputGrads[0]'
+      //   2. 'grad'
+      //   3. The captured grad in the callsite.
+      autogradContext_->accumulateGrad(
+          accumulateGrad_->variable,
+          inputGrads[0],
+          3 /* num_expected_refs */);
```

Comment:
Hope my understanding for `num_expected_refs` is correct here. The test case for non-copying grad does pass.
@pritamdamania87

Reply:
> 2. 'grad'

This should not bump the reference count of the Tensor since it is `const Tensor&` ?

> 3. The captured grad in the callsite.

Is this is referring to `captured_vars_` ?, if so lets mention `captured_vars_` here instead?

Reply:
Where is the third reference coming from? Aren't there 2 references? One in `captured_vars_ `and one in `inputGrads`?

Reply:
It's indeed 3. I printed `use_count()` in a test. Another one is `inputs` in the callsite. You are right that `grad` doesn't bump ref count. 

```
InputBuffer& inputs) {
```
Anyway, the ref count is quite fragile here. A callsite in the future can have an external reference (no other internal references) to the grad and this operator will incorrectly move(steal) a sparse tensor.

Reply:
nit: maybe add a comment here explaining where these come from so that we won't have to hunt them down next time.